### PR TITLE
[#86] Prune zero balance accounts

### LINKED ISSUE #86: Prune 0 balances
# Clarification and motivation

It seems that currently we don't remove addresses with 0 balance. That's inefficient because we unnecessarily store more data than necessary.

# Acceptance criteria

We store only positive balances. If balance of an address becomes 0, it's removed completely.

### PR DESCRIPTION
**Pull request by sras**
## Description

Change contract that when an account ends up with a zero balance, it is removed from the ledger. Add tests for transfer and burn to check zero balance account is removed after the operations.

## Related issue(s)

Resolves #86 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [x] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: haskell/test/Lorentz/Contracts/Test/Management.hs [hunk 1]
```diff
@@ -201,6 +201,28 @@ managementSpec originate = do
 
         lExpectViewConsumerStorage consumer [balanceExpected]
 
+    it "removes zero balance accounts" $ integrationalTestExpectation $ do
```
[the remark this answers is not in the document]
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Management.hs:204]: I put it in Management since despite transfer operation being specified in FA2, the specific behavior is not part FA2.

>AFAIU it is inside describe "Contract pausing". Why?

I think It was a mistake, will put it in separate section (or move to FA2 tests if it makes sense)

File: ligo/stablecoin/actions.ligo [hunk 2]
```diff
@@ -144,7 +144,9 @@ function debit_from
       then failwith ("FA2_INSUFFICIENT_BALANCE")
       else
       { const updated_balance : nat = abs (curr_balance - parameter.amount)
-      ; updated_ledger := Big_map.update(parameter.from_, Some (updated_balance), updated_ledger)
+      ; if updated_balance = 0n
+          then updated_ledger := Big_map.remove(parameter.from_, updated_ledger);
+          else updated_ledger := Big_map.update(parameter.from_, Some (updated_balance), updated_ledger);
```
[the remark this answers is not in the document]
Comment (sras) [reply ligo/stablecoin/actions.ligo]: @gromakovsky I have refactored that section a bit as you suggested.

File: haskell/test/Lorentz/Contracts/Test/Management.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Management.hs]: Moved to a new "describe" section for transfer

File: ligo/stablecoin/actions.ligo (hunk 2 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply ligo/stablecoin/actions.ligo]: Did you check that the contract became smaller? There were some cases when I expect a certain change to make the contract smaller, but it became larger :) So in this case I am not sure whether my suggested change is good or not. It may be good, that's why I suggested it, but it may be bad as well.

File: haskell/test/Lorentz/Contracts/Test/Management.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply haskell/test/Lorentz/Contracts/Test/Management.hs]: But "Management" in the module name means that this module tests management operations, i. e. those available only to managers (contract owner, master minter, pauser, etc.).
`Lorentz.Contracts.Test.FA2` tests this specific contract, not a general FA2 contract. For example, there is `validates token id` test which checks that `token_id = 1` is rejected. FA2 doesn't require `token_id = 1` to be rejected. And `OriginationParams` doesn't specify which token_ids are permitted. So I treat `Lorentz.Contracts.Test.FA2` as a module that tests entrypoints defined in FA2, keeping stablecoin specifics in mind.

File: ligo/stablecoin/actions.ligo (hunk 2 above)
[the remark this answers is not in the document]
Comment (sras) [reply ligo/stablecoin/actions.ligo]: @gromakovsky Did a simple lines of code comparison on generated contract now..

The contract size now is 1597 lines. In `master` it was 1600. And in this branch before refactoring, it was 1613 lines.

File: haskell/test/Lorentz/Contracts/Test/Management.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Management.hs]: I see. I ll move it then.

File: ligo/stablecoin/actions.ligo (hunk 2 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply ligo/stablecoin/actions.ligo]: Now that we have `stablecoin-client` it should be quite easy to originate the contract to carthagenet and see operation costs in block explorer. Can you do that please (for latest version and before refactoring)? You can download it from CI, then you won't need to build anything. Should take 5 minutes unless I am missing something.

[the remark this answers is not in the document]
Comment (sras) [reply ligo/stablecoin/actions.ligo]: Ah yes. Il do.

[the remark this answers is not in the document]
Comment (sras) [reply ligo/stablecoin/actions.ligo]: @gromakovsky Took more time that expected because I mixed up builds and ended up with same contract for different branches...anyway, here is the block explorer links..

Latest: https://carthage.tzkt.io/ooDPJ8nGyxNRS6egmSY9robVKSQ9Xw6srwGeQj5aMAHKsTzNags
Before refactoring : https://carthage.tzkt.io/op8y4mNhM3CJe8S4DYnnsr3YSnydXrgB2XWBRzzKNEM47SgEF6L
Master: https://carthage.tzkt.io/oorYmjnU33oyjwRvDedKdGM6hNh6U3cQs67f6EnW76NysQVz3TL

Gas costs:

Latest: 293945
Before refactor: 296763
Master : 293739

File: haskell/test/Lorentz/Contracts/Test/Management.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Management.hs]: Moved.

File: ligo/stablecoin/actions.ligo (hunk 2 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply ligo/stablecoin/actions.ligo]: Ok, cool, thanks.
